feat(*): various lemmas from the sensitivity project

robertylewis opened the pull request:
https://leanprover.zulipchat.com/#narrow/stream/116395-maths/topic/sensitivity.20project.20lemmas

Moving more over from https://github.com/leanprover-community/lean-sensitivity in preparation to put that in `archive`.

For most of these, I haven't thought much about naming, location, etc, so please make suggestions.

This has turned up an instance loop in `data.set.finite`. From Zulip:

> This has come up a few times (https://leanprover.zulipchat.com/#narrow/stream/113488-general/topic/fintype.2Ecard, https://leanprover.zulipchat.com/#narrow/stream/116395-maths/topic/.60decidable_mem_of_fintype.60.20problem): there's a type class loop. With `A : Type, fintype A, s : set A`, a search for `fintype ↥s` loops. `subtype.fintype` directs it to find `decidable_pred s`, and then `decidable_mem_of_fintype` directs it to find `fintype ↥s`.

I'm not sure of the best course of action here.

TO CONTRIBUTORS:

Make sure you have:

  * [ ] reviewed and applied the coding style: [coding](https://github.com/leanprover/mathlib/blob/master/docs/contribute/style.md), [naming](https://github.com/leanprover/mathlib/blob/master/docs/contribute/naming.md)
  * [ ] reviewed and applied [the documentation requirements](https://github.com/leanprover/mathlib/blob/master/docs/contribute/doc.md)
  * [ ] for tactics:
     * [ ] added or adapted documentation in [tactics.md](https://github.com/leanprover/mathlib/blob/master/docs/tactics.md)
     * [ ] write an example of use of the new feature in [tactics.lean](https://github.com/leanprover/mathlib/blob/master/test/tactics.lean)
  * [ ] make sure definitions and lemmas are put in the right files
  * [ ] make sure definitions and lemmas are not redundant

If this PR is related to a discussion on Zulip, please include a link in the discussion.

For reviewers: [code review check list](https://github.com/leanprover/mathlib/blob/master/docs/contribute/code-review.md)

robertylewis committed:
feat(*): various lemmas from the sensitivity project

jcommelin commented on src/data/set/finite.lean:
```diff
@@ -451,6 +451,20 @@ begin
+local attribute [instance, priority 1] classical.prop_decidable
+local attribute [instance, priority 0] set.decidable_mem_of_fintype
+
+lemma to_finset_card {α : Type*} [fintype α] (H : set α) :
```
Should this be a simp-lemma? Which of the two sides is simp-normal-form?

robertylewis replied:
I don't know, which is why I didn't mark it as one. They have the same depth (because of the coercion on the rhs).

jcommelin commented on src/linear_algebra/dual.lean:
```diff
@@ -206,3 +207,118 @@ def eval_equiv (h : dim K V < omega) : V ≃ₗ[K] dual K (dual K V) :=
+universes u v w
+variables {K : Type u} {V : Type v} {ι : Type w} [decidable_eq ι]
+variables [discrete_field K] [add_comm_group V] [vector_space K V]
+          [decidable_eq V] [decidable_eq (ι → V)] [decidable_eq $ dual K V]
```
Are these `decidable_eq`s still necessary?

robertylewis replied:
I've linted the file and removed them from a few declarations.

jcommelin commented on src/linear_algebra/dual.lean:
```diff
@@ -206,3 +207,118 @@ def eval_equiv (h : dim K V < omega) : V ≃ₗ[K] dual K (dual K V) :=
+
+include h
+
+lemma dual_lc (l : ι →₀ K) (i : ι) : ε i (dual_pair.lc e l) = l i :=
```
```suggestion
@[simp] lemma dual_lc (l : ι →₀ K) (i : ι) : ε i (dual_pair.lc e l) = l i :=
```

robertylewis replied:
Lean reports that `dual_lc` is an invalid simp lemma. I don't know why.

kim-em replied:
Perhaps it is upset about the "side condition" `(h : dual_pair e ε)`? I'm not sure either.

jcommelin reviewed (commented):
There's a bunch of `decidable_eq`s and I'm not up to speed where they are necessary and where not. Did you lint `dual.lean`?

robertylewis committed:
fix proof broken by nonterminal simp

robertylewis committed:
Update src/linear_algebra/dual.lean

Co-Authored-By: Johan Commelin <johan@commelin.net>

robertylewis committed:
lint dual.lean

robertylewis committed:
remove decidable_mem_of_fintype instance

this leads to loops with `subtype.fintype` under the right decidable_eq assumptions

robertylewis committed:
dual_lc is invalid simp lemma

robertylewis committed:
fix namespace

robertylewis committed:
add extra lemma

robertylewis commented on src/ring_theory/power_series.lean:
```diff
@@ -164,14 +164,14 @@ end
 ext $ λ n,
 begin
-  rw [coeff_mul, finset.sum_eq_single (n, (0 : σ →₀ ℕ))];
-  simp [mem_antidiagonal_support, coeff_one],
```
This is a good example of why nonterminal `simp`s are bad. This `simp` was killing two of three goals, and making progress on the third.

ChrisHughes24 commented on src/algebra/module.lean:
```diff
@@ -420,3 +420,17 @@ def is_add_group_hom.to_linear_map [add_comm_group α] [add_comm_group β]
+    rw [add_smul, add_smul, one_smul, ih, one_smul] }
+end
+
+@[simp] lemma finset.sum_const' {α : Type*} {R : Type*} [ring R] {β : Type*}
```
I don't think this should be `@[simp]`, there's not way to infer `R` from the left hand side. Also, maybe `R` should be explicit since there's no way to infer it.

robertylewis replied:
Oh, good catch, thanks. This was only being used for a backward rewrite in the final proof.

ChrisHughes24 commented on src/linear_algebra/dual.lean:
```diff
@@ -206,3 +209,122 @@ def eval_equiv (h : dim K V < omega) : V ≃ₗ[K] dual K (dual K V) :=
+universes u v w
+variables {K : Type u} {V : Type v} {ι : Type w} [dι : decidable_eq ι]
+variables [discrete_field K] [add_comm_group V] [vector_space K V]
+          [decidable_eq V] [dιv : decidable_eq (ι → V)] [decidable_eq $ dual K V]
```
Are these `decidable_eq` assumptions still necessary after the `lc` refactor?

robertylewis replied:
This is still way more finicky than I would like. Short answer, no, they're not necessary. But it's nontrivial to remove them. I've done it (see the coming commit), but it took some trial and error, and I expect it will have an effect on the sensitivity proof (which I haven't checked against these changes yet).

The fact that they were unnecessary was also not caught by `sanity_check`; they're unnecessary, but not unused.

robertylewis replied:
Good news: removing these had no effect on the sensitivity proof.

ChrisHughes24 commented on src/linear_algebra/dual.lean:
```diff
@@ -206,3 +209,122 @@ def eval_equiv (h : dim K V < omega) : V ≃ₗ[K] dual K (dual K V) :=
+omit h
+
+@[simp]
+lemma coeffs_apply (v : V) (i : ι) : h.coeffs v i = ε i v := rfl
```
`@[simp]` on same line as statement

ChrisHughes24 commented on src/linear_algebra/dual.lean:
```diff
@@ -206,3 +209,122 @@ def eval_equiv (h : dim K V < omega) : V ≃ₗ[K] dual K (dual K V) :=
+local notation `V'` := dual K V
+
+/-- `e` and `ε` have characteristic properties of a basis and its dual -/
+structure dual_pair (e : ι → V) (ε : ι → V') :=
```
An explanation of this could be added to the header as well

ChrisHughes24 commented on src/linear_algebra/finsupp.lean:
```diff
@@ -420,3 +420,17 @@ begin
+
+lemma linear_map.map_finsupp_total {R : Type*} {β : Type*} {γ : Type*} [ring R] [decidable_eq R]
+  [add_comm_group β] [decidable_eq β] [module R β]
+  [add_comm_group γ] [decidable_eq γ] [module R γ]
```
Are these `decidable_eq` assumptions still necessary?

robertylewis replied:
Nope, and these were easier to remove.

robertylewis committed:
fix sum_const

robertylewis committed:
remove unnecessary dec_eq assumptions

robertylewis committed:
remove decidable_eq assumptions

robertylewis committed:
document dual.lean

robertylewis reviewed (commented):
I've added documentation to `dual.lean` and removed more unnecessary `decidable_eq` assumptions.

jcommelin commented on src/linear_algebra/finsupp.lean:
```diff
@@ -8,12 +8,13 @@ Linear structures on function with finite support `α →₀ M`.
 
 namespace finsupp
 
+local attribute [instance, priority 10] classical.prop_decidable
```
Maybe I'm missing something. Why don't you simply write `open_locale classical`?

robertylewis replied:
Because I just moved this from a few lines above. Changed.

robertylewis committed:
use classical locale

Commit:
remove some unnecessary includes

Commit:
remove an unused variable

kim-em commented on src/linear_algebra/dual.lean:
```diff
@@ -53,7 +88,7 @@ lemma to_dual_apply (i j : ι) :
 
-omit h
+omit de h
 def eval_finsupp_at (i : ι) : (ι →₀ K) →ₗ[K] K :=
```
```suggestion
/-- Evaluation of finitely supported functions at a fixed point `i`, as a `K`-linear map. -/
def eval_finsupp_at (i : ι) : (ι →₀ K) →ₗ[K] K :=
```

kim-em committed:
Update src/linear_algebra/dual.lean

Commit:
fixing a doc comment

Commit:
Merge branch 'sensitivity' of github.com:leanprover-community/mathlib into sensitivity

ChrisHughes24 committed:
Merge branch 'master' into sensitivity
